### PR TITLE
Add support for constructors in Python modules

### PR DESCRIPTION
**Pull request by helderco**
Follow-up to #6135

Equivalent example in Python:
```python
from dagger.mod import object_type, field, function

@object_type
class Test:
    foo: str = field(default="defaultfoo")

    @function
    def echo(self) -> str:
        return self.foo
```
```shell
❯ dagger call echo
defaultfoo
❯ dagger call --foo=abc echo
abc
```

### CODE REVIEW

**Comment by helderco** [comment]
I'm going to make some changes:
- Collect the actual constructor method instead of the class callable to get a more accurate signature
- Wrap object instantiation in a worker thread to allow for running async tasks in the constructor in the main thread

**Comment by helderco** [comment]
Ready for review.

**Comment by gerhard** [comment]
Waiting on this one to get merged so that the following security items will get fixed:
- https://github.com/dagger/dagger/security/dependabot/393
- https://github.com/dagger/dagger/security/dependabot/394

File: core/integration/module_test.go [hunk 1]
```diff
@@ -3000,6 +3132,23 @@ func daggerFunctions(args ...string) dagger.WithContainerFunc {
 	}
 }
 
+func sdkSource(sdk, contents string) dagger.WithContainerFunc {
```
Comment (sipsma) [inline core/integration/module_test.go:3135]: Nice 😎

File: sdk/python/src/dagger/mod/_converter.py [hunk 2]
```diff
@@ -59,24 +61,29 @@ def dagger_type_unstructure(obj):
     return conv
 
 
+@functools.cache
```
Comment (sipsma) [inline sdk/python/src/dagger/mod/_converter.py:64]: Mostly just curious, how does this differentiate different inputs for `annotation: type`? i.e. does it consider each instance of the object to be distinct vs. does it do some sort of "deep equals"-like logic on the value of the object?

**Review by sipsma — APPROVED** [review]

File: sdk/python/src/dagger/mod/_converter.py (hunk 2 above)
Comment (helderco) [reply sdk/python/src/dagger/mod/_converter.py:64]: It's based on [hash](https://docs.python.org/3/library/functions.html?highlight=hash#hash):
```python
>>> to_typedef(str | None)
<dagger.TypeDef object at 0x1029032c0>
>>> to_typedef(None | str)
<dagger.TypeDef object at 0x1029032c0>
>>> to_typedef.cache_info()
CacheInfo(hits=1, misses=1, maxsize=None, currsize=1)
>>> hash(str | None)
9088350826229485722
>>> hash(None | str)
9088350826229485722
```
The argument `annotation` should only be types and they are stable and deterministic. If for some reason it isn't, it'll just cache miss but still execute.

**Pull request merged**